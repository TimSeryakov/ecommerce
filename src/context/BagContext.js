// import { useAsyncStorage } from '@react-native-community/async-storage';
import propTypes from 'prop-types';
import React, { useState, useMemo, useCallback } from 'react';

const BagContext = React.createContext();

function BagProvider({ children }) {
  const [bag, setBag] = useState([]);

  const addProduct = useCallback(product => {
    setBag(bag => [...bag, product]);
  }, []);

  const removeProduct = useCallback(id => {
    setBag(bag => bag.filter(product => product.id !== id));
  }, []);

  const bagSize = useMemo(() => {
    return bag.length;
  }, [bag]);

  const defaultValue = useMemo(
    () => ({
      bag,
      bagSize,
      addProduct,
      removeProduct,
    }),
    [bag, bagSize, addProduct, removeProduct]
  );
  return <BagContext.Provider value={defaultValue}>{children}</BagContext.Provider>;
}

BagProvider.propTypes = {
  children: propTypes.oneOfType([propTypes.node.isRequired, propTypes.arrayOf(propTypes.node)])
    .isRequired,
};

function useBag() {
  const context = React.useContext(BagContext);
  if (context === undefined) {
    throw new Error(`useBag must be used within a CustomerProvider`);
  }
  return context;
}

export { BagProvider, useBag };
