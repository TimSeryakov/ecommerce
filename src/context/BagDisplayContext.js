import propTypes from 'prop-types';
import React, { useState, useContext, useCallback, useMemo } from 'react';

const BagDisplayContext = React.createContext();
const BagControllerContext = React.createContext();

function BagDisplayProvider({ children }) {
  const [isBagOpen, setBagOpen] = useState(false);

  const showBag = useCallback(() => {
    setBagOpen(true);
  }, []);

  const closeBag = useCallback(() => {
    setBagOpen(false);
  }, []);

  const controllers = useMemo(
    () => ({
      showBag,
      closeBag,
    }),
    [showBag, closeBag]
  );

  return (
    <BagControllerContext.Provider value={controllers}>
      <BagDisplayContext.Provider value={isBagOpen}>{children}</BagDisplayContext.Provider>
    </BagControllerContext.Provider>
  );
}

BagDisplayProvider.propTypes = {
  children: propTypes.oneOfType([propTypes.node.isRequired, propTypes.arrayOf(propTypes.node)])
    .isRequired,
};

function useBagDisplay() {
  const context = useContext(BagDisplayContext);
  if (context === undefined) {
    throw new Error(`useBagDisplay must be used within a BagDisplayProvider`);
  }
  return context;
}

function useBagController() {
  const context = useContext(BagControllerContext);
  if (context === undefined) {
    throw new Error(`useBagController must be used within a BagControllerProvider`);
  }
  return context;
}

export {
  BagDisplayProvider,
  useBagDisplay,
  useBagController,
  BagControllerContext,
  BagDisplayContext,
};
