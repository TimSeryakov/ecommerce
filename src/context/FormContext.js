import { useFormik } from 'formik';
import propTypes from 'prop-types';
import React, { createContext, useContext, useMemo, useCallback } from 'react';

const FormContext = createContext();

const FormContextProvider = ({ initialValues, validationSchema, onSubmit, children }) => {
  const {
    handleChange,
    handleBlur,
    values,
    touched,
    errors,
    handleSubmit,
    setFieldValue,
    setFieldError,
    setFieldTouched,
    submitForm,
    setValues,
    unregisterField: removeField,
  } = useFormik({
    initialValues,
    validationSchema,
    onSubmit,
    validateOnChange: true,
    validateOnMount: true,
  });

  const registerField = useCallback(
    field => {
      setFieldValue(field, '');
    },
    [setFieldValue]
  );

  const unregisterField = useCallback(
    field => {
      setFieldError(field, 'Campo Removido');
      setFieldValue(field, undefined);
      removeField(field);
    },
    [setFieldError, setFieldValue, removeField]
  );

  const cleanFieldError = useCallback(
    field => {
      setFieldError(field, undefined);
    },
    [setFieldError]
  );

  const validateField = useCallback(
    async field => {
      try {
        await validationSchema.validateAt(field, values);
        return true;
      } catch (err) {
        setFieldError(field, err.errors[0]);
        setFieldTouched(field, true, true);
        return false;
      }
    },
    [values, setFieldError, setFieldTouched, validationSchema]
  );

  const defaultContext = useMemo(
    () => ({
      handleChange,
      handleBlur,
      values,
      touched,
      errors,
      handleSubmit,
      setFieldValue,
      setFieldError,
      setFieldTouched,
      registerField,
      setValues,
      unregisterField,
      validateField,
      cleanFieldError,
      submitForm,
    }),
    [
      handleChange,
      handleBlur,
      values,
      touched,
      errors,
      handleSubmit,
      setFieldValue,
      setValues,
      setFieldError,
      setFieldTouched,
      registerField,
      unregisterField,
      validateField,
      cleanFieldError,
      submitForm,
    ]
  );

  return <FormContext.Provider value={defaultContext}>{children}</FormContext.Provider>;
};

FormContextProvider.propTypes = {
  initialValues: propTypes.shape().isRequired,
  validationSchema: propTypes.shape(),
  onSubmit: propTypes.func.isRequired,
  children: propTypes.oneOfType([
    propTypes.element.isRequired,
    propTypes.arrayOf(propTypes.element.isRequired),
  ]).isRequired,
};

FormContextProvider.defaultProps = {
  validationSchema: null,
};

function useForm() {
  const context = useContext(FormContext);
  if (context === undefined) {
    throw new Error(`useForm must be used within a FormProvider`);
  }
  return context;
}

const useField = field => {
  const form = useContext(FormContext);

  return {
    handleChange: form.handleChange(field),
    handleBlur: form.handleBlur(field),
    value: form.values[field],
    values: form.values,
    touched: form.touched[field],
    error: form.errors[field],
    handleSubmit: form.handleSubmit,
    setFieldValue: form.setFieldValue,
    setFieldTouched: form.setFieldTouched,
    registerField: form.registerField,
    unregisterField: form.unregisterField,
    setFieldError: form.setFieldError,
  };
};

export { FormContext, FormContextProvider, useForm, useField };
