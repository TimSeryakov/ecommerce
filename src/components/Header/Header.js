import { Link } from 'gatsby';
import propTypes from 'prop-types';
import React, { useState } from 'react';

import BagIcon from '../../assets/svg/BagIcon';
import MenuIcon from '../../assets/svg/MenuIcon';
import { useBag } from '../../context/BagContext';

const HeaderLink = ({ children, path }) => {
  return (
    <Link
      to={path}
      className="cursor-pointer block p-4 font-semibold text-base leading-thight tracking-wide text-black hover:opacity-50 sm:p-0 sm:m-0 sm:border-none sm:mx-4"
    >
      {children}
    </Link>
  );
};

HeaderLink.propTypes = {
  children: propTypes.string.isRequired,
  path: propTypes.string.isRequired,
};

const Header = ({ showBag }) => {
  const [open, setOpen] = useState(false);
  const { bagSize } = useBag();

  return (
    <header className="flex flex-row items-baseline justify-between px-6 py-6 mb-12 border-solid border-b border-gray-300">
      <div className="w-full flex flex-col items-start sm:items-baseline sm:justify-between sm:flex-row">
        <div className="flex items-center">
          <button
            aria-label="Menu"
            type="button"
            onClick={() => setOpen(!open)}
            className="cursor-pointer text-black h-6 w-6 mr-4 sm:hidden focus:outline-none focus:shadow-outline"
          >
            <MenuIcon open={open} />
          </button>
          <Link to="/">
            <h1 className="text-3xl text-black">Plants.</h1>
          </Link>
        </div>

        <div
          className={`${
            open ? 'flex' : 'hidden'
          } w-full flex-col mt-4 sm:my-0 sm:w-auto sm:flex sm:flex-row sm:items-center`}
        >
          <HeaderLink path="/">All</HeaderLink>
          <HeaderLink path="/new">New</HeaderLink>
          <HeaderLink path="/on-sale">On Sale</HeaderLink>
          <button
            aria-label="Show Bag"
            onClick={showBag}
            type="button"
            className="relative cursor-pointer hidden text-black h-6 w-6 mr-4 mb-1 focus:outline-none focus:shadow-outline sm:block sm:ml-8"
          >
            <BagIcon />
            <div className="absolute right-0 top-0 -mr-3 -mt-2 bg-black rounded-full h-5 w-5 flex items-center justify-center">
              <span className="text-white font-semibold text-2xs">{bagSize}</span>
            </div>
          </button>
        </div>
      </div>

      <button
        aria-label="Show Bag"
        onClick={showBag}
        type="button"
        className="relative cursor-pointer text-black h-6 w-6 mr-4 mb-1 focus:outline-none focus:shadow-outline sm:hidden sm:ml-8"
      >
        <BagIcon />
        <div className="absolute right-0 top-0 -mr-3 -mt-2 bg-black rounded-full h-5 w-5 flex items-center justify-center">
          <span className="text-white font-semibold text-2xs">{bagSize}</span>
        </div>
      </button>
    </header>
  );
};

Header.propTypes = {
  showBag: propTypes.func.isRequired,
};

export default Header;
