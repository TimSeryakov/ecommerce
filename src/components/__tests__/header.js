import React from 'react';
import renderer from 'react-test-renderer';

import { BagProvider } from '../../context/BagContext';
import Header from '../Header';

describe('Header', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <BagProvider>
          <Header />
        </BagProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
