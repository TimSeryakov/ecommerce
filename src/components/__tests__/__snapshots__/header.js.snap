// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<header
  className="flex flex-row items-baseline justify-between px-6 py-6 bg-white border-solid border-b border-gray-300"
>
  <div
    className="w-full flex flex-col items-start sm:items-baseline sm:justify-between sm:flex-row"
  >
    <div
      className="flex items-center"
    >
      <button
        className="cursor-pointer text-black h-6 w-6 mr-4 sm:hidden focus:outline-none"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="stroke-current"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 6h16M4 12h16M4 18h16"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
      </button>
      <a
        href="/"
      >
        <h1
          className="text-3xl font-thin text-black"
        >
          Plants.
        </h1>
      </a>
    </div>
    <div
      className="hidden w-full flex-col mt-4 sm:my-0 sm:w-auto sm:flex sm:flex-row sm:items-center"
    >
      <a
        className="cursor-pointer block p-4 font-semibold text-base leading-thight tracking-wide text-black hover:opacity-50 sm:p-0 sm:m-0 sm:border-none sm:mx-4"
        href="/shop/"
      >
        All
      </a>
      <a
        className="cursor-pointer block p-4 font-semibold text-base leading-thight tracking-wide text-black hover:opacity-50 sm:p-0 sm:m-0 sm:border-none sm:mx-4"
        href="/"
      >
        New
      </a>
      <a
        className="cursor-pointer block p-4 font-semibold text-base leading-thight tracking-wide text-black hover:opacity-50 sm:p-0 sm:m-0 sm:border-none sm:mx-4"
        href="/"
      >
        On Sale
      </a>
      <button
        className="relative cursor-pointer hidden text-black h-6 w-6 mr-4 mb-1 focus:outline-none sm:block sm:ml-8"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="stroke-current"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
        >
          <path
            d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
          />
        </svg>
        <div
          className="absolute right-0 top-0 -mr-3 -mt-2 bg-black rounded-full h-5 w-5 flex items-center justify-center"
        >
          <span
            className="text-white font-semibold text-2xs"
          >
            0
          </span>
        </div>
      </button>
    </div>
  </div>
  <button
    className="relative cursor-pointer text-black h-6 w-6 mr-4 mb-1 focus:outline-none sm:hidden sm:ml-8"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="stroke-current"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
    >
      <path
        d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
      />
    </svg>
    <div
      className="absolute right-0 top-0 -mr-3 -mt-2 bg-black rounded-full h-5 w-5 flex items-center justify-center"
    >
      <span
        className="text-white font-semibold text-2xs"
      >
        0
      </span>
    </div>
  </button>
</header>
`;
